nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java

```diff
@@ -0,0 +1,216 @@
+        if (sslContextService != null) {
+            sslContext = sslContextService.createSSLContext(SSLContextService.ClientAuth.REQUIRED);
+        }
+
+        // if we decide to support SSL then get the context and pass it in here
+        return new SocketChannelDispatcher<>(eventFactory, handlerFactory, bufferPool, events,
+            getLogger(), maxConnections, sslContext, charSet);
+    }
+
+
+    @Override
+    protected String getBatchKey(BeatsEvent event) {
+        return event.getSender();
+    }
+
+    protected void respond(final BeatsEvent event, final BeatsResponse beatsResponse) {
+        final ChannelResponse response = new BeatsChannelResponse(beatsEncoder, beatsResponse);
+
+        final ChannelResponder responder = event.getResponder();
+        responder.addResponse(response);
+        try {
+            responder.respond();
+        } catch (IOException e) {
+            getLogger().error("Error sending response for transaction {} due to {}",
+                new Object[]{event.getSeqNumber(), e.getMessage()}, e);
+        }
+    }
+
+    protected void postProcess(final ProcessContext context, final ProcessSession session, final List<BeatsEvent> events) {
+        // first commit the session so we guarantee we have all the events successfully
+        // written to FlowFiles and transferred to the success relationship
+        session.commit();
+        // respond to each event to acknowledge successful receipt
+        for (final BeatsEvent event : events) {
+            respond(event, BeatsResponse.ok(event.getSeqNumber()));
+        }
+    }
+
+    @Override
+    protected String getTransitUri(FlowFileEventBatch batch) {
+        final String sender = batch.getEvents().get(0).getSender();
+        final String senderHost = sender.startsWith("/") && sender.length() > 1 ? sender.substring(1) : sender;
+        final String transitUri = new StringBuilder().append("beats").append("://").append(senderHost).append(":")
+            .append(port).toString();
+        return transitUri;
+    }
+
+    @Override
+    protected Map<String, String> getAttributes(FlowFileEventBatch batch) {
+        final List<BeatsEvent> events = batch.getEvents();
+        // the sender and command will be the same for all events based on the batch key
+        final String sender = events.get(0).getSender();
+        final int numAttributes = events.size() == 1 ? 5 : 4;
+        final Map<String, String> attributes = new HashMap<>(numAttributes);
+        attributes.put(beatsAttributes.SENDER.key(), sender);
+        attributes.put(beatsAttributes.PORT.key(), String.valueOf(port));
+        attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+        // if there was only one event then we can pass on the transaction
+        // NOTE: we could pass on all the transaction ids joined together
```

Comment:
wait a second. it is already there:

@WritesAttribute(attribute = "beats.sequencenumber", description = "The sequence number of the message. Only included if <Batch Size> is 1."),